node/src/components/storage.rs

```diff
@@ -1422,38 +1695,302 @@ impl Storage {
+fn get_single_block_body_v2<Tx: Transaction>(
+    tx: &mut Tx,
+    block_body_v2_db: Database,
+    deploy_hashes_db: Database,
+    transfer_hashes_db: Database,
+    proposer_db: Database,
+    block_body_hash: &Digest,
+) -> Result<Option<BlockBody>, LmdbExtError> {
+    let deploy_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+        match get_merkle_linked_list_node(tx, block_body_v2_db, deploy_hashes_db, block_body_hash)?
+        {
+            Some(deploy_hashes_with_proof) => deploy_hashes_with_proof,
+            None => return Ok(None),
+        };
+    let transfer_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+        match get_merkle_linked_list_node(
+            tx,
+            block_body_v2_db,
+            transfer_hashes_db,
+            deploy_hashes_with_proof.merkle_proof_of_rest(),
+        )? {
+            Some(transfer_hashes_with_proof) => transfer_hashes_with_proof,
+            None => return Ok(None),
+        };
+    let proposer_with_proof: MerkleLinkedListNode<PublicKey> = match get_merkle_linked_list_node(
+        tx,
+        block_body_v2_db,
+        proposer_db,
+        transfer_hashes_with_proof.merkle_proof_of_rest(),
+    )? {
+        Some(proposer_with_proof) => {
+            debug_assert_eq!(*proposer_with_proof.merkle_proof_of_rest(), hash::SENTINEL1);
+            proposer_with_proof
+        }
+        None => return Ok(None),
+    };
+    let block_body = BlockBody::new(
+        proposer_with_proof.take_value(),
+        deploy_hashes_with_proof.take_value(),
+        transfer_hashes_with_proof.take_value(),
+    );
+
+    Ok(Some(block_body))
+}
+
+// TODO: remove once we run the garbage collection again
+#[allow(dead_code)]
+fn garbage_collect_block_body_v2_db(
+    txn: &mut RwTransaction,
+    block_body_v2_db: &Database,
+    deploy_hashes_db: &Database,
+    transfer_hashes_db: &Database,
+    proposer_db: &Database,
+    block_body_hash_to_header_map: &BTreeMap<Digest, BlockHeader>,
+) -> Result<(), Error> {
+    // This will store all the keys that are reachable from a block header, in all the parts
+    // databases (we're basically doing a mark-and-sweep below).
+    // The entries correspond to: the block_body_v2_db, deploy_hashes_db, transfer_hashes_db,
+    // proposer_db respectively.
+    let mut live_digests: [HashSet<Digest>; BlockBody::PARTS_COUNT + 1] = [
+        HashSet::new(),
+        HashSet::new(),
+        HashSet::new(),
+        HashSet::new(),
+    ];
+
+    for (body_hash, header) in block_body_hash_to_header_map {
+        if header.hashing_algorithm_version() != HashingAlgorithmVersion::V2 {
+            // We're only interested in body hashes for V2 blocks here
             continue;
         }
+        let mut current_digest = *body_hash;
+        let mut live_digests_index = 1;
+        while current_digest != hash::SENTINEL1 && !live_digests[0].contains(&current_digest) {
```

Comment:
Should we actually also stop here if:

```rust
live_digests[live_digests_index].contains(&current_digest)
```

?

Reply:
I think no, because `current_digest` should only be a key in the `block_body_v2_db`, and not in whichever database the respective part belongs to. And I'm wary of shortcuts like this across databases, because it's what caused a bug before (a hash of an empty array could be reachable, say, in the `deploy_hashes_db` but not in `transfer_hashes_db`, but since we weren't keeping any info regarding which database a key refers to, that was causing that key to _not_ be dropped from `transfer_hashes_db`, even though it should have been).

Reply:
Ah, you're right, I got confused! :facepalm: